File: docs/geometry.js
// Pure-JS mesh construction primitives for a round, tapered nursery pot.
// Direct port of geometry.py. Triangles are represented as arrays of
// three [x,y,z] points; winding conventions were empirically verified in
// the Python version via signed-volume tests against analytical frustum
// volumes — see geometry.py comments for the derivations.

const TWO_PI = 2 * Math.PI;

export function circleXY(radius, n, cx = 0, cy = 0) {
  const pts = [];
  for (let i = 0; i < n; i++) {
    const a = (TWO_PI * i) / n;
    pts.push([cx + radius * Math.cos(a), cy + radius * Math.sin(a)]);
  }
  return pts;
}

export function ring3(radius, z, n, cx = 0, cy = 0) {
  return circleXY(radius, n, cx, cy).map(([x, y]) => [x, y, z]);
}

// outward=true => normal points away from the shared axis (e.g. outer pot
// wall); outward=false => normal points toward the axis (inner wall / hole
// wall). Valid for rings that differ in z (lateral/wall surfaces).
export function quadStrip(ringA, ringB, outward = true) {
  const n = ringA.length;
  const tris = [];
  for (let i = 0; i < n; i++) {
    const j = (i + 1) % n;
    const a0 = ringA[i], a1 = ringA[j];
    const b0 = ringB[i], b1 = ringB[j];
    if (outward) {
      tris.push([a0, a1, b1]);
      tris.push([a0, b1, b0]);
    } else {
      tris.push([a0, b1, a1]);
      tris.push([a0, b0, b1]);
    }
  }
  return tris;
}

export function flatDiscFan(radius, z, n, cx = 0, cy = 0, facingUp = true) {
  const ring = ring3(radius, z, n, cx, cy);
  const center = [cx, cy, z];
  const tris = [];
  for (let i = 0; i < n; i++) {
    const j = (i + 1) % n;
    if (facingUp) tris.push([center, ring[i], ring[j]]);
    else tris.push([center, ring[j], ring[i]]);
  }
  return tris;
}

// Flat washer cap (e.g. pot's top rim). Note: for a flat ring at constant z
// with increasing radius, quadStrip's outward=true actually yields a -z
// (down) normal — its convention was derived for radius-vs-z lateral
// walls, not radius-vs-radius flats — so this intentionally inverts.
export function annulusCap(rOuter, rInner, z, n, facingUp = true) {
  const outer = ring3(rOuter, z, n);
  const inner = ring3(rInner, z, n);
  return quadStrip(inner, outer, !facingUp);
}

// ---------------------------------------------------------------------
// Ear-clipping triangulation for a polygon with circular holes (2D)
// ---------------------------------------------------------------------

function signedArea(pts) {
  let s = 0;
  for (let i = 0; i < pts.length; i++) {
    const [x0, y0] = pts[i];
    const [x1, y1] = pts[(i + 1) % pts.length];
    s += x0 * y1 - x1 * y0;
  }
  return 0.5 * s;
}

function crossZ(o, a, b) {
  return (a[0] - o[0]) * (b[1] - o[1]) - (a[1] - o[1]) * (b[0] - o[0]);
}

function pointInTri(p, a, b, c, eps) {
  const d1 = crossZ(a, b, p);
  const d2 = crossZ(b, c, p);
  const d3 = crossZ(c, a, p);
  const hasNeg = d1 < -eps || d2 < -eps || d3 < -eps;
  const hasPos = d1 > eps || d2 > eps || d3 > eps;
  return !(hasNeg && hasPos);
}

function dist2d(p, q) {
  return Math.hypot(p[0] - q[0], p[1] - q[1]);
}

export function earClip(polyPtsIn) {
  const pts = polyPtsIn.slice();
  let idx = pts.map((_, i) => i);
  if (signedArea(pts) < 0) idx = idx.reverse();

  const xs = pts.map((p) => p[0]);
  const ys = pts.map((p) => p[1]);
  const span = Math.max(Math.max(...xs) - Math.min(...xs), Math.max(...ys) - Math.min(...ys), 1.0);
  const epsArea = span * span * 1e-9;
  const epsDup = span * 1e-7;

  const triangles = [];
  const maxGuard = idx.length * idx.length + 100;
  let guard = 0;
  while (idx.length > 3 && guard < maxGuard) {
    guard++;
    const n = idx.length;
    let found = false;
    for (let k = 0; k < n; k++) {
      const i0 = idx[(k - 1 + n) % n];
      const i1 = idx[k];
      const i2 = idx[(k + 1) % n];
      const a = pts[i0], b = pts[i1], c = pts[i2];
      if (crossZ(a, b, c) <= epsArea) continue;
      let ok = true;
      for (const m of idx) {
        if (m === i0 || m === i1 || m === i2) continue;
        const pm = pts[m];
        if (dist2d(pm, a) < epsDup || dist2d(pm, b) < epsDup || dist2d(pm, c) < epsDup) continue;
        if (pointInTri(pm, a, b, c, epsArea)) {
          ok = false;
          break;
        }
      }
      if (ok) {
        triangles.push([i0, i1, i2]);
        idx.splice(k, 1);
        found = true;
        break;
      }
    }
    if (!found) break;
  }
  if (idx.length === 3) {
    const [i0, i1, i2] = idx;
    if (crossZ(pts[i0], pts[i1], pts[i2]) > 0) triangles.push([i0, i1, i2]);
  }
  return { triangles, pts };
}

function bridgeHole(boundary, hole) {
  let best = Infinity, bi = 0, bj = 0;
  for (let i = 0; i < boundary.length; i++) {
    for (let j = 0; j < hole.length; j++) {
      const d = dist2d(boundary[i], hole[j]);
      if (d < best) {
        best = d;
        bi = i;
        bj = j;
      }
    }
  }
  return [
    ...boundary.slice(0, bi + 1),
    ...hole.slice(bj),
    ...hole.slice(0, bj + 1),
    ...boundary.slice(bi),
  ];
}

export function discWithHoles2D(outerRadius, holeCenters, holeRadius, nOuter = 48, nHole = 14) {
  let boundary = circleXY(outerRadius, nOuter);
  for (const [hx, hy] of holeCenters) {
    const hole = circleXY(holeRadius, nHole, hx, hy).reverse(); // CW hole
    boundary = bridgeHole(boundary, hole);
  }
  return earClip(boundary);
}

export function discWithHoles3D(outerRadius, holeCenters, holeRadius, z, nOuter = 48, nHole = 14, facingUp = true) {
  const { triangles, pts } = discWithHoles2D(outerRadius, holeCenters, holeRadius, nOuter, nHole);
  const out = [];
  for (const [i0, i1, i2] of triangles) {
    const p0 = [pts[i0][0], pts[i0][1], z];
    const p1 = [pts[i1][0], pts[i1][1], z];
    const p2 = [pts[i2][0], pts[i2][1], z];
    out.push(facingUp ? [p0, p1, p2] : [p0, p2, p1]);
  }
  return out;
}

export function holeTunnelWalls(holeCenters, holeRadius, z0, z1, nHole = 14) {
  const out = [];
  for (const [hx, hy] of holeCenters) {
    const ringA = ring3(holeRadius, z0, nHole, hx, hy);
    const ringB = ring3(holeRadius, z1, nHole, hx, hy);
    out.push(...quadStrip(ringA, ringB, false));
  }
  return out;
}

export function cylinderSolid(radius, z0, z1, n = 24, cx = 0, cy = 0, topRadius = null) {
  const r0 = radius;
  const r1 = topRadius === null ? radius : topRadius;
  const ringA = ring3(r0, z0, n, cx, cy);
  const ringB = ring3(r1, z1, n, cx, cy);
  const side = quadStrip(ringA, ringB, true);
  const bottom = flatDiscFan(r0, z0, n, cx, cy, false);
  const top = flatDiscFan(r1, z1, n, cx, cy, true);
  return [...side, ...bottom, ...top];
}

export function signedVolume(triangles) {
  let v = 0;
  for (const [p0, p1, p2] of triangles) {
    const cx = p1[1] * p2[2] - p1[2] * p2[1];
    const cy = p1[2] * p2[0] - p1[0] * p2[2];
    const cz = p1[0] * p2[1] - p1[1] * p2[0];
    v += p0[0] * cx + p0[1] * cy + p0[2] * cz;
  }
  return v / 6.0;
}

File: geometry.py
"""Pure-numpy mesh construction primitives for a round, tapered nursery pot.

No CAD/boolean library is available in this environment, so the model is
built as a single watertight solid via careful ring/cap construction
(equivalent to a lathe "revolve" profile) rather than boolean subtraction.
Drainage holes are built directly into the floor caps (ear-clipped polygon
with holes) rather than cut afterward. Feet are separate small watertight
solids that sit flush against the underside — standard practice for
script-generated printable parts (no true CSG union needed since the
faces only touch, they don't interpenetrate).

All angles in radians internally; degrees accepted at the public API edges.
"""
import numpy as np

TWO_PI = 2 * np.pi


# ---------------------------------------------------------------------------
# Basic point/ring helpers
# ---------------------------------------------------------------------------

def circle_xy(radius, n, cx=0.0, cy=0.0):
    """CCW ordered (n,2) points around a circle, angle 0 at +x axis."""
    ang = np.linspace(0, TWO_PI, n, endpoint=False)
    x = cx + radius * np.cos(ang)
    y = cy + radius * np.sin(ang)
    return np.stack([x, y], axis=1)


def ring3(radius, z, n, cx=0.0, cy=0.0):
    xy = circle_xy(radius, n, cx, cy)
    z_col = np.full((n, 1), z)
    return np.concatenate([xy, z_col], axis=1)


def quad_strip(ring_a, ring_b, outward=True):
    """Lateral (side) surface triangles connecting two same-length rings.
    outward=True => normals point away from the shared axis (e.g. outer
    pot wall); outward=False => normals point toward the axis (e.g. inner
    pot wall or a hole's own wall)."""
    n = len(ring_a)
    tris = []
    for i in range(n):
        j = (i + 1) % n
        a0, a1 = ring_a[i], ring_a[j]
        b0, b1 = ring_b[i], ring_b[j]
        if outward:
            tris.append([a0, a1, b1])
            tris.append([a0, b1, b0])
        else:
            tris.append([a0, b1, a1])
            tris.append([a0, b0, b1])
    return np.array(tris)


def flat_disc_fan(radius, z, n, cx=0.0, cy=0.0, facing_up=True):
    """Simple fan-triangulated flat disc (no holes)."""
    ring = ring3(radius, z, n, cx, cy)
    center = np.array([cx, cy, z])
    tris = []
    for i in range(n):
        j = (i + 1) % n
        if facing_up:
            tris.append([center, ring[i], ring[j]])
        else:
            tris.append([center, ring[j], ring[i]])
    return np.array(tris)


def annulus_cap(r_outer, r_inner, z, n, facing_up=True):
    """Flat washer-shaped cap (e.g. the pot's top rim edge). Note: for a
    flat ring at constant z with increasing radius, quad_strip's
    outward=True actually yields a -z (down) normal (its winding
    convention was derived for radius-vs-z lateral walls, not
    radius-vs-radius flats) — so this is intentionally inverted relative
    to quad_strip's usual meaning. Verified against signed_volume tests."""
    outer = ring3(r_outer, z, n)
    inner = ring3(r_inner, z, n)
    return quad_strip(inner, outer, outward=(not facing_up))


# ---------------------------------------------------------------------------
# Ear-clipping triangulation for a polygon with circular holes (2D)
# ---------------------------------------------------------------------------

def _signed_area(pts):
    x = pts[:, 0]
    y = pts[:, 1]
    return 0.5 * np.sum(x * np.roll(y, -1) - np.roll(x, -1) * y)


def _cross_z(o, a, b):
    return (a[0] - o[0]) * (b[1] - o[1]) - (a[1] - o[1]) * (b[0] - o[0])


def _point_in_tri(p, a, b, c, eps):
    d1 = _cross_z(a, b, p)
    d2 = _cross_z(b, c, p)
    d3 = _cross_z(c, a, p)
    has_neg = (d1 < -eps) or (d2 < -eps) or (d3 < -eps)
    has_pos = (d1 > eps) or (d2 > eps) or (d3 > eps)
    return not (has_neg and has_pos)


def ear_clip(poly_pts):
    """poly_pts: (n,2) array, any orientation. Returns list of index
    triples (into poly_pts) forming a triangulation. Tolerant of the
    coincident duplicate vertices produced by hole-bridging."""
    pts = np.asarray(poly_pts, dtype=np.float64)
    idx = list(range(len(pts)))
    if _signed_area(pts) < 0:
        idx = idx[::-1]

    span = max(np.ptp(pts[:, 0]), np.ptp(pts[:, 1]), 1.0)
    eps_area = span * span * 1e-9
    eps_dup = span * 1e-7

    triangles = []
    guard = 0
    max_guard = len(idx) ** 2 + 100
    while len(idx) > 3 and guard < max_guard:
        guard += 1
        n = len(idx)
        found = False
        for k in range(n):
            i0, i1, i2 = idx[(k - 1) % n], idx[k], idx[(k + 1) % n]
            a, b, c = pts[i0], pts[i1], pts[i2]
            if _cross_z(a, b, c) <= eps_area:
                continue  # reflex or degenerate
            ok = True
            for m in idx:
                if m in (i0, i1, i2):
                    continue
                pm = pts[m]
                # skip points coincident with a triangle vertex (bridge dupes)
                if (np.linalg.norm(pm - a) < eps_dup or
                        np.linalg.norm(pm - b) < eps_dup or
                        np.linalg.norm(pm - c) < eps_dup):
                    continue
                if _point_in_tri(pm, a, b, c, eps_area):
                    ok = False
                    break
            if ok:
                triangles.append((i0, i1, i2))
                idx.pop(k)
                found = True
                break
        if not found:
            break  # numerical fallback: leave remainder untriangulated
    if len(idx) == 3:
        a, b, c = pts[idx[0]], pts[idx[1]], pts[idx[2]]
        if _cross_z(a, b, c) > 0:
            triangles.append((idx[0], idx[1], idx[2]))
    return triangles, pts


def _bridge_hole(boundary, hole):
    """Merge one hole polygon into the boundary polygon via nearest-vertex
    bridge (standard technique for turning polygon-with-hole into a
    simple polygon before ear clipping)."""
    b = np.asarray(boundary)
    h = np.asarray(hole)
    d = np.linalg.norm(b[:, None, :] - h[None, :, :], axis=2)
    i, j = np.unravel_index(np.argmin(d), d.shape)
    new_boundary = (
        list(boundary[: i + 1])
        + list(hole[j:])
        + list(hole[: j + 1])
        + list(boundary[i:])
    )
    return new_boundary


def disc_with_holes_2d(outer_radius, hole_centers, hole_radius, n_outer=48, n_hole=14):
    """Returns (triangles_idx, pts) for a disc of outer_radius centered at
    origin, minus circular holes at hole_centers (list of (x,y)) with
    hole_radius. Triangulation only — caller lifts to 3D + orients."""
    outer = list(map(tuple, circle_xy(outer_radius, n_outer)))
    boundary = outer
    for (hx, hy) in hole_centers:
        hole = list(map(tuple, circle_xy(hole_radius, n_hole, hx, hy)))
        hole = hole[::-1]  # CW so it reads as a hole against CCW boundary
        boundary = _bridge_hole(boundary, hole)
    tri_idx, pts = ear_clip(np.array(boundary))
    return tri_idx, pts


def disc_with_holes_3d(outer_radius, hole_centers, hole_radius, z, n_outer=48, n_hole=14, facing_up=True):
    tri_idx, pts = disc_with_holes_2d(outer_radius, hole_centers, hole_radius, n_outer, n_hole)
    tris = []
    for (i0, i1, i2) in tri_idx:
        p0 = np.array([pts[i0][0], pts[i0][1], z])
        p1 = np.array([pts[i1][0], pts[i1][1], z])
        p2 = np.array([pts[i2][0], pts[i2][1], z])
        if facing_up:
            tris.append([p0, p1, p2])
        else:
            tris.append([p0, p2, p1])
    return np.array(tris)


def hole_tunnel_walls(hole_centers, hole_radius, z0, z1, n_hole=14):
    """Vertical tube wall for each drainage hole, normal pointing inward
    (into the void), connecting the bottom cap opening to the top-of-floor
    cap opening."""
    all_tris = []
    for (hx, hy) in hole_centers:
        ring_a = ring3(hole_radius, z0, n_hole, hx, hy)
        ring_b = ring3(hole_radius, z1, n_hole, hx, hy)
        all_tris.append(quad_strip(ring_a, ring_b, outward=False))
    return np.concatenate(all_tris, axis=0)


def cylinder_solid(radius, z0, z1, n=24, cx=0.0, cy=0.0, top_radius=None):
    """Standalone watertight cylinder (or frustum if top_radius given) —
    used for feet. z1 > z0."""
    r0 = radius
    r1 = radius if top_radius is None else top_radius
    ring_a = ring3(r0, z0, n, cx, cy)
    ring_b = ring3(r1, z1, n, cx, cy)
    side = quad_strip(ring_a, ring_b, outward=True)
    bottom = flat_disc_fan(r0, z0, n, cx, cy, facing_up=False)
    top = flat_disc_fan(r1, z1, n, cx, cy, facing_up=True)
    return np.concatenate([side, bottom, top], axis=0)


def signed_volume(triangles):
    """Sum of signed tetrahedron volumes (divergence theorem). Positive
    for a closed mesh with consistent outward-facing normals."""
    v0 = triangles[:, 0, :]
    v1 = triangles[:, 1, :]
    v2 = triangles[:, 2, :]
    return np.sum(np.einsum("ij,ij->i", v0, np.cross(v1, v2))) / 6.0
